descriptor/qsarmolecular/src/main/java/org/openscience/cdk/qsar/descriptors/molecular/SpiroAtomCountDescriptor.java

```diff
@@ -93,37 +90,64 @@ public void setParameters(Object[] params) throws CDKException {
         return NAMES;
     }
 
+    private static void traverseRings(IAtomContainer mol, IAtom atom, IBond prev) {
+        atom.setFlag(CDKConstants.VISITED, true);
+        prev.setFlag(CDKConstants.VISITED, true);
+        for (IBond bond : mol.getConnectedBondsList(atom)) {
+            IAtom nbr = bond.getOther(atom);
+            if (!nbr.getFlag(CDKConstants.VISITED))
+                traverseRings(mol, nbr, bond);
+            else
+                bond.setFlag(CDKConstants.VISITED, true);
+        }
+    }
+
+    private static int getSpiroDegree(IAtomContainer mol, IAtom atom) {
+        if (!atom.isInRing())
+            return 0;
+        List<IBond> rbonds = new ArrayList<>(4);
+        for (IBond bond : mol.getConnectedBondsList(atom)) {
+            if (bond.isInRing())
+                rbonds.add(bond);
+        }
+        if (rbonds.size() < 4)
+            return 0;
+        int degree = 0;
+        // clear flags
+        for (IBond b : mol.bonds())
+            b.setFlag(CDKConstants.VISITED, false);
+        for (IAtom a : mol.atoms())
+            a.setFlag(CDKConstants.VISITED, false);
+        // visit rings
+        atom.setFlag(CDKConstants.VISITED, true);
+        for (IBond rbond : rbonds) {
+            if (!rbond.getFlag(CDKConstants.VISITED)) {
+                traverseRings(mol, rbond.getOther(atom), rbond);
+                degree++;
+            }
+        }
+        return degree < 2 ? 0 : degree;
+    }
+
     /**
      * {@inheritDoc}
      */
     @Override
     public DescriptorValue calculate(IAtomContainer atomContainer) {
-
-        RingSearch rs = new RingSearch(atomContainer);
-        Set<Integer> spiroAtomIdx = new HashSet<>();
-
-        // rings identified this way are either spiro or have no common edges
-        int[][] isorings = rs.isolated();
-
-        for (int i = 0; i < isorings.length - 1; i++) {
-            for (int j = i + 1; j < isorings.length; j++) {
-                List<Integer> idxi = new ArrayList<>();
-                for (int k = 0; k < isorings[i].length; k++) idxi.add(isorings[i][k]);
-
-                List<Integer> idxj = new ArrayList<>();
-                for (int k = 0; k < isorings[j].length; k++) idxj.add(isorings[j][k]);
-
-                // find atom indices common to the two rings
-                idxi.retainAll(idxj);
-                
-                // for spiro rings, should just have 1 atom
-                if (idxi.size() == 1)
-                    spiroAtomIdx.add(idxi.get(0));
+        int nSpiro = 0;
+
+        try {
+            IAtomContainer local = atomContainer.clone();
```

Comment:
Would try and avoid cloning... but that's a deeper problem with the QSAR stuff.

Reply:
We're not modify the molecule here beyond setting the ring and visit flags. Since the rings should be invariant it should not matter to other descriptors if they are set or not.

Reply:
I agree - but without a clone this will modify the input molecule (and a test will fail), and I don't see a good way to reset to the original state. But even if that were possible, it would likely mess up threaded code using this class (and most other descriptor classes)

Reply:
I agree that the extent of modification is small (and likely not to impact other descriptors), but it is possible that calling code has some assumptions on the molecule state. I very much like immutable state, but clearly, it leads to inefficient code with the current architecture

Reply:
Yeah I figured that might be case, I don't like that but alas it is what is. No probs just leave the clone in.